File: src/Domain/Services/RightsDescriber.php
<?php

declare(strict_types=1);
declare(ticks=1000);

namespace App\Domain\Services;

use App\Domain\Factory\RightsDescribeFactory;
use App\Domain\ValueObject\RightsDescribe;

readonly class RightsDescriber
{
    public function __construct(
        private RightsDescribeFactory $rightsDescribeFactory,
    ) {
    }

    /**
     * @param string[] $roles
     */
    public function byRoles(array $roles): RightsDescribe
    {
        return $this->rightsDescribeFactory->byRoles($roles);
    }
}

File: internal/domain/user/rights_describer.go
package user

// RightsDescriber is a thin domain service that wraps the factory so
// consumers (e.g. GetMe handler) don't need to instantiate the factory
// themselves and the dependency stays explicit in DI.
type RightsDescriber struct {
	factory *RightsDescribeFactory
}

// NewRightsDescriber constructs the service.
func NewRightsDescriber(f *RightsDescribeFactory) *RightsDescriber {
	return &RightsDescriber{factory: f}
}

// ByRoles delegates to the underlying factory; kept on a separate type so
// future audit/logging hooks can attach without changing the factory.
func (s *RightsDescriber) ByRoles(roles []string) RightsDescribe {
	return s.factory.ByRoles(roles)
}
